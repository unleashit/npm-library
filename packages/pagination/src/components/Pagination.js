import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import throttle from "lodash/throttle";
import style from "./scss/pagination.scss";
import ChevronLeft from "./icons/chevron-left.svg";
import ChevronRight from "./icons/chevron-right.svg";

const isClient = typeof window !== "undefined";

class Pagination extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      containerWidth: 0
    };

    this.containerRef = React.createRef();
    this.boundSetContainerWidth = throttle(
      this.setContainerWidth.bind(this),
      500
    );
  }

  componentDidMount() {
    this.setContainerWidth();
    window.addEventListener("resize", this.boundSetContainerWidth);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.boundSetContainerWidth);
  }

  setContainerWidth() {
    this.setState({ containerWidth: this.containerRef.current.offsetWidth });
  }

  clickHandler(type, pageNumber = 1) {
    const { currentOffset, total, perPage, paginationHandler } = this.props;
    let newOffset;

    switch (type) {
      case "next":
        newOffset = currentOffset + perPage;
        if (newOffset > total) return false;
        paginationHandler(newOffset);
        break;
      case "prev":
        newOffset = currentOffset - perPage;
        if (newOffset < 0) return false;
        paginationHandler(newOffset);
        break;
      case "page":
        newOffset = (pageNumber - 1) * perPage;
        if (newOffset === currentOffset || newOffset > total || newOffset < 0)
          return false;
        paginationHandler(newOffset);
        break;
      default:
        return undefined;
    }

    return new TypeError(
      'Must supply pagination an argument of "prev", "next", or "page"'
    );
  }

  prev() {
    const { currentOffset, perPage } = this.props;

    return currentOffset - perPage >= 0 ? (
      <div className={style.prev} onClick={() => this.clickHandler("prev")}>
        <ChevronLeft className={style.chevronLeft} /> previous
      </div>
    ) : null;
  }

  next() {
    const { currentOffset, perPage, total } = this.props;

    return currentOffset + perPage < total ? (
      <div className={style.next} onClick={() => this.clickHandler("next")}>
        next <ChevronRight className={style.chevronRight} />
      </div>
    ) : null;
  }

  numbers() {
    const { total, perPage, currentOffset } = this.props;
    const { containerWidth } = this.state;
    console.log(containerWidth);

    const pages = Math.ceil(total / perPage);
    const currentPage = Math.floor(currentOffset / perPage) + 1;

    let maxPages;
    if (containerWidth <= 1000) {
      maxPages = [[450, 2], [550, 3], [650, 4], [850, 8], [1000, 10]].find(
        group => containerWidth < group[0]
      )[1];
    } else {
      maxPages = containerWidth / 80;
    }

    console.log("max pages: ", maxPages);
    let pageAry = new Array(pages).fill(undefined).map((_, i) => i + 1);

    if (pages > maxPages) {
      if (currentPage > 1 && currentPage < pages) {
        pageAry = pageAry.slice(
          currentPage - 1,
          currentPage - 1 + (maxPages - 1)
        );
        if (pageAry[pageAry.length - 1] === pages) pageAry.pop();
        pageAry.unshift(1, "...");
        pageAry.push("...", pages);
      } else if (currentPage === 1) {
        pageAry = pageAry.slice(0, maxPages);
        pageAry.push("...", pages);
      } else if (currentPage === pages) {
        pageAry = pageAry.slice(-maxPages);
        pageAry.unshift(1, "...");
      }
    }

    return pages > 1
      ? pageAry.map(
          (page, i) =>
            page === "..." ? (
              <span key={`${i}-ellipsis`} className={style.ellipsis}>
                ...
              </span>
            ) : (
              <span
                key={`${page}-pagination`}
                className={classnames(style.number, {
                  active: page === currentPage
                })}
                onClick={() => this.clickHandler("page", page)}
              >
                {page}
              </span>
            )
        )
      : null;
  }

  render() {
    const { total, perPage } = this.props;

    return total > perPage ? (
      <div
        className={classnames(style.paginationContainer)}
        ref={this.containerRef}
      >
        <div className={style.pagination}>
          {this.prev()}
          {this.numbers()}
          {this.next()}
        </div>
      </div>
    ) : null;
  }
}

Pagination.propTypes = {
  currentOffset: PropTypes.number.isRequired,
  perPage: PropTypes.number,
  paginationHandler: PropTypes.func.isRequired,
  total: PropTypes.number.isRequired
};

Pagination.defaultProps = {
  currentOffset: 0,
  perPage: 10,
  total: 0
};

export default Pagination;
