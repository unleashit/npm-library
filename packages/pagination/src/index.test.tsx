/* eslint-disable @typescript-eslint/explicit-function-return-type */
import * as React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallow } from 'enzyme';
import Pagination from './index';

// declare const shallow: any;

describe('<Pagination />', () => {
  let wrapper: any;
  let props: any = {};

  jest
    .spyOn(Pagination.prototype, 'setContainerWidth')
    .mockImplementation(function setContainerWidth(this: { setState: any }) {
      this.setState({ containerWidth: 1200 });
    });

  beforeEach(() => {
    props = {
      currentOffset: 0,
      perPage: 10,
      paginationHandler: jest.fn(),
      total: 50,
    };

    wrapper = shallow(<Pagination {...props} />);
  });

  it('renders without crashing', () => {
    expect(wrapper.find('.unl-pagination__container')).toHaveLength(1);
    const { containerWidth } = wrapper.state();
    expect(containerWidth).toEqual(1200);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders next button correctly', () => {
    expect(wrapper.find('.unl-pagination__inner').text()).toContain('next');
    expect(wrapper.find('.unl-pagination__inner').text()).not.toContain('prev');
  });

  it('renders prev button correctly', () => {
    props.currentOffset = 50;
    wrapper = shallow(<Pagination {...props} />);

    expect(wrapper.find('.unl-pagination__inner').text()).toContain('prev');
    expect(wrapper.find('.unl-pagination__inner').text()).not.toContain('next');
  });

  it('renders both prev and next btns correctly', () => {
    props.currentOffset = 25;
    wrapper = shallow(<Pagination {...props} />);

    expect(wrapper.find('.unl-pagination__inner').text()).toContain('prev');
    expect(wrapper.find('.unl-pagination__inner').text()).toContain('next');
  });

  it('renders the right amount of pages', () => {
    const tests = [
      { perPage: 5, total: 20, ex: 4 },
      { perPage: 5, total: 21, ex: 5 },
      { perPage: 5, total: 16, ex: 4 },
      { perPage: 5, total: 15, ex: 3 },
      { perPage: 10, total: 100, ex: 10 },
      { perPage: 10, total: 98, ex: 10 },
      { perPage: 12, total: 50, ex: 5 },
    ];

    tests.forEach(({ perPage, total, ex }) => {
      props.perPage = perPage;
      props.total = total;
      wrapper = shallow(<Pagination {...props} />);

      expect(
        wrapper.find('.unl-pagination__inner > .unl-pagination__number'),
      ).toHaveLength(ex);
    });
  });

  describe('click events', () => {
    it('#next btn calls the pagination handler with the right offset', () => {
      let nextBtn = wrapper.find('.unl-pagination__next');
      nextBtn.simulate('click');
      expect(props.paginationHandler).toHaveBeenLastCalledWith(10);

      props.currentOffset = 20;
      props.perPage = 5;
      wrapper = shallow(<Pagination {...props} />);
      nextBtn = wrapper.find('.unl-pagination__next');
      nextBtn.simulate('click');
      expect(props.paginationHandler).toHaveBeenLastCalledWith(25);
    });

    it('#prev btn calls the pagination handler with the right offset', () => {
      props.currentOffset = 30;
      props.perPage = 10;
      wrapper = shallow(<Pagination {...props} />);
      const prevBtn = wrapper.find('.unl-pagination__prev');
      prevBtn.simulate('click');
      expect(props.paginationHandler).toHaveBeenLastCalledWith(20);
    });

    it('#page number btns call the pagination handler with the right offset', () => {
      const tests = [
        { currentOffset: 15, perPage: 10, pageClicked: 1, ex: 0 },
        { currentOffset: 15, perPage: 10, pageClicked: 4, ex: 30 },
        { currentOffset: 0, perPage: 10, pageClicked: 1 }, // calling current page
        { currentOffset: 0, perPage: 20, pageClicked: 3, ex: 40 },
        { currentOffset: 37, perPage: 5, pageClicked: 5, ex: 20 },
        { currentOffset: 0, perPage: 5, pageClicked: 5, ex: 20 },
        { currentOffset: 45, perPage: 5, pageClicked: 10 }, // calling current page
        { currentOffset: 45, perPage: 5, pageClicked: 2, ex: 5 },
      ];

      tests.forEach(({ currentOffset, perPage, pageClicked, ex }) => {
        props = { ...props, currentOffset, perPage, pageClicked };
        props.paginationHandler.mockClear();

        wrapper = shallow(<Pagination {...props} />);
        const firstPageBtn = wrapper.find('.unl-pagination__number').at(pageClicked - 1);

        firstPageBtn.simulate('click');

        if (pageClicked - 1 === currentOffset / perPage) {
          // not called when clicking on current page
          expect(props.paginationHandler).not.toHaveBeenCalled();
        } else {
          expect(props.paginationHandler).toHaveBeenLastCalledWith(ex);
        }
      });
    });
  });
});
