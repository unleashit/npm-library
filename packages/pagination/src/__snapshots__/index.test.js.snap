// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders without crashing 1`] = `
<div
  className="pagination__container undefined"
>
  <div
    className="pagination__inner undefined"
  >
    <span
      className="pagination__number undefined active"
      key="1-pagination"
      onClick={[Function]}
    >
      1
    </span>
    <span
      className="pagination__number undefined"
      key="2-pagination"
      onClick={[Function]}
    >
      2
    </span>
    <span
      className="pagination__number undefined"
      key="3-pagination"
      onClick={[Function]}
    >
      3
    </span>
    <span
      className="pagination__number undefined"
      key="4-pagination"
      onClick={[Function]}
    >
      4
    </span>
    <span
      className="pagination__number undefined"
      key="5-pagination"
      onClick={[Function]}
    >
      5
    </span>
    <div
      className="pagination__next undefined"
      onClick={[Function]}
    >
      next 
      <test-file-stub />
    </div>
  </div>
</div>
`;
